### ISSUE TITLE
[tune][Bug] RayActorError (worker crashed) when resources_per_trial is not set

### ISSUE THREAD

**Issue by Numchies** [ray-project/ray#23109]
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Tune

### Issue Severity

High: It blocks me to complete my task.

### What happened + What you expected to happen

Hi all,
I'm quite new when it comes to programming, so please bear with me if I don't understand your explanations right away or if I ask you to ELI5. I'm currently forecasting energy consumption for my Master's thesis using an LSTM.
**1. Describe the bug**
I'm trying to use Ray Tuner to perform a gridsearch for the hyperparameters of my LSTM while using multiple cores. The code (file run in Anaconda Prompt) works _fine_ for 15-60 mins and then it gives me an error after failing two trials (models):
```
2022-03-12 17:16:39,372 WARNING util.py:163 -- The `start_trial` operation took 0.642 s, which may be a performance bottleneck.
 pid=44452) Epoch 1/50
2022-03-12 17:16:41,164 WARNING worker.py:1326 -- Traceback (most recent call last):
  File "python\ray\_raylet.pyx", line 633, in ray._raylet.execute_task
  File "python\ray\_raylet.pyx", line 674, in ray._raylet.execute_task
  File "python\ray\_raylet.pyx", line 640, in ray._raylet.execute_task
  File "python\ray\_raylet.pyx", line 644, in ray._raylet.execute_task
  File "python\ray\_raylet.pyx", line 593, in ray._raylet.execute_task.function_executor
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\_private\function_manager.py", line 648, in actor_method_executor
    return method(__ray_actor, *args, **kwargs)
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\_private\function_manager.py", line 544, in temporary_actor_method
    raise RuntimeError(
RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:

Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\python\pywrap_tensorflow.py", line 60, in <module>
    from tensorflow.python._pywrap_tensorflow_internal import *
ImportError: DLL load failed while importing _pywrap_tensorflow_internal: Die Auslagerungsdatei ist zu klein, um diesen Vorgang durchzuführen.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\_private\function_manager.py", line 603, in _load_actor_class_from_gcs
    actor_class = pickle.loads(pickled_class)
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\cloudpickle\cloudpickle.py", line 679, in subimport
    __import__(name)
  File "C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\__init__.py", line 37, in <module>
    from tensorflow.python.tools import module_util as _module_util
  File "C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\python\__init__.py", line 36, in <module>
    from tensorflow.python import pywrap_tensorflow as _pywrap_tensorflow
  File "C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\python\pywrap_tensorflow.py", line 75, in <module>
    raise ImportError(
ImportError: Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\python\pywrap_tensorflow.py", line 60, in <module>
    from tensorflow.python._pywrap_tensorflow_internal import *
ImportError: DLL load failed while importing _pywrap_tensorflow_internal: Die Auslagerungsdatei ist zu klein, um diesen Vorgang durchzuführen.


Failed to load the native TensorFlow runtime.
See https://www.tensorflow.org/install/errors for some common causes and solutions.
If you need help, create an issue at https://github.com/tensorflow/tensorflow/issues and include the entire stack trace above this error message.


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "python\ray\_raylet.pyx", line 774, in ray._raylet.task_execution_handler
  File "python\ray\_raylet.pyx", line 595, in ray._raylet.execute_task
  File "python\ray\_raylet.pyx", line 715, in ray._raylet.execute_task
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\_private\function_manager.py", line 544, in temporary_actor_method
    raise RuntimeError(
RuntimeError: The actor with name ImplicitFunc failed to import on the worker. This may be because needed library dependencies are not installed in the worker environment:

Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\python\pywrap_tensorflow.py", line 60, in <module>
    from tensorflow.python._pywrap_tensorflow_internal import *
ImportError: DLL load failed while importing _pywrap_tensorflow_internal: Die Auslagerungsdatei ist zu klein, um diesen Vorgang durchzuführen.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\_private\function_manager.py", line 603, in _load_actor_class_from_gcs
    actor_class = pickle.loads(pickled_class)
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\cloudpickle\cloudpickle.py", line 679, in subimport
    __import__(name)
  File "C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\__init__.py", line 37, in <module>
    from tensorflow.python.tools import module_util as _module_util
  File "C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\python\__init__.py", line 36, in <module>
    from tensorflow.python import pywrap_tensorflow as _pywrap_tensorflow
  File "C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\python\pywrap_tensorflow.py", line 75, in <module>
    raise ImportError(
ImportError: Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\python\pywrap_tensorflow.py", line 60, in <module>
    from tensorflow.python._pywrap_tensorflow_internal import *
ImportError: DLL load failed while importing _pywrap_tensorflow_internal: Die Auslagerungsdatei ist zu klein, um diesen Vorgang durchzuführen.


Failed to load the native TensorFlow runtime.
See https://www.tensorflow.org/install/errors for some common causes and solutions.
If you need help, create an issue at https://github.com/tensorflow/tensorflow/issues and include the entire stack trace above this error message.

An unexpected internal error occurred while the worker was executing a task.
2022-03-12 17:16:41,194 WARNING worker.py:1326 -- A worker died or was killed while executing a task by an unexpected system error. To troubleshoot the problem, check the logs for the dead worker. RayTask ID: ffffffffffffffffefefc240fa0478483d69566801000000 Worker ID: 5bfc1a7ca7eae5d8f0de7be96a059a6024ce1c4ea02ec9e5534a5902 Node ID: 02857c7d18a93c1422ec3dacf75dfa881fb0984633ee806426305790 Worker IP address: 127.0.0.1 Worker port: 53620 Worker PID: 23692`

These messages continue for a few lines until:
`2022-03-12 17:17:24,522 ERROR import_thread.py:85 -- ImportThread: <_MultiThreadedRendezvous of RPC that terminated with:
        status = StatusCode.UNKNOWN
        details = "Stream removed"
        debug_error_string = "{"created":"@1647101844.519000000","description":"Error received from peer ipv4:127.0.0.1:57599","file":"src/core/lib/surface/call.cc","file_line":1075,"grpc_message":"Stream removed","grpc_status":2}"
>
Exception in thread ray_listen_error_messages:
Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\lib\threading.py", line 973, in _bootstrap_inner
    self.run()
  File "C:\ProgramData\Anaconda3\lib\threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\worker.py", line 1311, in listen_error_messages_from_gcs
    _, error_data = worker.gcs_error_subscriber.poll()
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\_private\gcs_pubsub.py", line 340, in poll
    self._poll_locked(timeout=timeout)
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\_private\gcs_pubsub.py", line 270, in _poll_locked
    fut.result(timeout=1)
  File "C:\ProgramData\Anaconda3\lib\site-packages\grpc\_channel.py", line 744, in result
    raise self
grpc._channel._MultiThreadedRendezvous: <_MultiThreadedRendezvous of RPC that terminated with:
        status = StatusCode.UNKNOWN
        details = "Stream removed"
        debug_error_string = "{"created":"@1647101844.519000000","description":"Error received from peer ipv4:127.0.0.1:57599","file":"src/core/lib/surface/call.cc","file_line":1075,"grpc_message":"Stream removed","grpc_status":2}"
>
Exception in thread ray_print_logs:
Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\lib\threading.py", line 973, in _bootstrap_inner
    self.run()
  File "C:\ProgramData\Anaconda3\lib\threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\worker.py", line 464, in print_logs
    msg = subscriber.poll()
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\_private\gcs_pubsub.py", line 374, in poll
    self._poll_locked(timeout=timeout)
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\_private\gcs_pubsub.py", line 270, in _poll_locked
    fut.result(timeout=1)
  File "C:\ProgramData\Anaconda3\lib\site-packages\grpc\_channel.py", line 744, in result
    raise self
grpc._channel._MultiThreadedRendezvous: <_MultiThreadedRendezvous of RPC that terminated with:
        status = StatusCode.UNKNOWN
        details = "Stream removed"
        debug_error_string = "{"created":"@1647101844.519000000","description":"Error received from peer ipv4:127.0.0.1:57599","file":"src/core/lib/surface/call.cc","file_line":1075,"grpc_message":"Stream removed","grpc_status":2}"
>
2022-03-12 17:17:40,185 ERROR ray_trial_executor.py:558 -- Trial train_model_d1260_00066: Unexpected error starting runner.
Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\tune\ray_trial_executor.py", line 549, in start_trial
    return self._start_trial(trial)
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\tune\ray_trial_executor.py", line 447, in _start_trial
    runner = self._setup_remote_runner(trial)
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\tune\ray_trial_executor.py", line 305, in _setup_remote_runner
    trainable_cls = trial.get_trainable_cls()
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\tune\trial.py", line 684, in get_trainable_cls
    return get_trainable_cls(self.trainable_name)
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\tune\registry.py", line 35, in get_trainable_cls
    return _global_registry.get(TRAINABLE_CLASS, trainable_name)
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\tune\registry.py", line 163, in get
    value = _internal_kv_get(_make_key(self._prefix, category, key))
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\_private\client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\experimental\internal_kv.py", line 45, in _internal_kv_get
    return global_gcs_client.internal_kv_get(key, namespace)
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\_private\gcs_utils.py", line 144, in wrapper
    return f(self, *args, **kwargs)
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\_private\gcs_utils.py", line 230, in internal_kv_get
    reply = self._kv_stub.InternalKVGet(req)
  File "C:\ProgramData\Anaconda3\lib\site-packages\grpc\_channel.py", line 946, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "C:\ProgramData\Anaconda3\lib\site-packages\grpc\_channel.py", line 849, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
        status = StatusCode.UNAVAILABLE
        details = "failed to connect to all addresses"
        debug_error_string = "{"created":"@1647101860.186000000","description":"Failed to pick subchannel","file":"src/core/ext/filters/client_channel/client_channel.cc","file_line":3135,"referenced_errors":[{"created":"@1647101860.186000000","description":"failed to connect to all addresses","file":"src/core/lib/transport/error_utils.cc","file_line":163,"grpc_status":14}]}"
>
2022-03-12 17:17:42,209 WARNING util.py:163 -- The `start_trial` operation took 17.863 s, which may be a performance bottleneck.
[2022-03-12 17:18:25,142 C 34944 34244] gcs_client.cc:343: Couldn't reconnect to GCS server. The last attempted GCS server address was 127.0.0.1:57599
*** StackTrace Information ***

Windows fatal exception: access violation
```

If I check the error files of the failed trials they say:
```
Failure # 1 (occurred at 2022-03-12_17-17-09)
Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\tune\trial_runner.py", line 886, in _process_trial
    results = self.trial_executor.fetch_result(trial)
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\tune\ray_trial_executor.py", line 675, in fetch_result
    result = ray.get(trial_future[0], timeout=DEFAULT_GET_TIMEOUT)
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\_private\client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "C:\ProgramData\Anaconda3\lib\site-packages\ray\worker.py", line 1765, in get
    raise value
ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.
	class_name: ImplicitFunc
	actor_id: 0c052c4627e702e05222ad4501000000
	pid: 74936
	namespace: 89964be2-a188-4cb7-9300-a07a100b64dd
	ip: 127.0.0.1
The actor is dead because its worker process has died. Worker exit type: SYSTEM_ERROR_EXIT
```

**2. Describe expected behavior**
I'd really like to use Ray Tune for my hyperparameter optimization and would have expected the program to finish the execution of the gridsearch without any problems. Sadly, I couldn't find a lot of information on the "RayActorError" and don't know how to fix this issue by myself. I'd greatly appreciate your support.

**3. Describe useful information (e.g., logs)**
See above.

Thanks already in advance for your help. I very much appreciate it!

### Versions / Dependencies

Python 3.9.7
ray 1.11.0
tensorflow 2.8.0

### Reproduction script

```
# 0. Import Libraries
import pandas as pd
import numpy as np
import matplotlib as mpl
#import matplotlib.pyplot as plt
#import plotly.express as px
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from math import sqrt
#from tensorflow import keras
import tensorflow as tf
from tensorflow.keras.preprocessing.sequence import TimeseriesGenerator
from sklearn.preprocessing import MinMaxScaler, StandardScaler
import ray
from ray import tune
from ray.tune.integration.keras import TuneReportCallback
import os

os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'

np.set_printoptions(linewidth=1000,precision=5)
pd.set_option('display.max.columns', 15)
mpl.rcParams['figure.figsize']=(17, 8)
mpl.rcParams['axes.grid']=True
    
# 1. Read data and choose input parameters
df=pd.read_csv('Example_XXX.csv', sep=',')
df['timestamp_utc'] = pd.to_datetime(df['timestamp_utc'], 

                                     infer_datetime_format=True)
print("ORIGINAL DF:")
for i in range(0,len(df.columns)):
    print(str(i)+": "+df.columns[i])
    i+=1
print("\t")

df_input = df.drop(df.columns[[0]], axis=1) 

print("DF COLUMNS USED FOR MODELLING:")
for i in range(0,len(df_input.columns)):
    print(str(i)+": "+df_input.columns[i])
    i+=1
print("\t")
    
data=df_input.to_numpy()

features=data[:,1:]
target=data[:,0]
        
# 2. Scale data and create generators
win_length=720 # how many data points are input
batch_size=16 # how many data points are output per 'batch' // should only be power of 2
num_features=features.shape[1]

train_set, test_set=train_test_split(data, test_size=0.15, shuffle=False)

scaler=StandardScaler()
train_scaled=scaler.fit_transform(train_set)

x_train=train_scaled[:,1:]
y_train=train_scaled[:,0]

x_train, x_validate, y_train, y_validate=train_test_split(
    x_train, y_train, test_size=0.176, shuffle=False)

test_scaled=scaler.fit_transform(test_set)

x_test=test_scaled[:,1:]
y_test=test_scaled[:,0]

train_generator=TimeseriesGenerator(x_train, y_train, length=win_length, 
                                      sampling_rate=1, 
                                      batch_size=batch_size)
validate_generator=TimeseriesGenerator(x_validate, y_validate, 
                                       length=win_length, sampling_rate=1,
                                       batch_size=batch_size)
test_generator=TimeseriesGenerator(x_test, y_test, 
                                        length=win_length, 
                                        sampling_rate=1, 
                                        batch_size=batch_size)

#%%
# 3. Define an objective function.
def train_model(config):
    
    n_nodes=config['n_nodes']
    n_layers=config['n_layers']
    learning_rate=config['learning_rate']
    dropout=config['dropout']
    
    #unpack config
    model=tf.keras.Sequential()
    i=0
    while i < n_layers:
        if i==0:
            model.add(tf.keras.layers.LSTM(n_nodes, input_shape=(win_length, num_features),
                                           return_sequences=True))
        else:
            model.add(tf.keras.layers.LSTM(n_nodes,return_sequences=True))
        i+=1
    if i==0:
        model.add(tf.keras.layers.LSTM(n_nodes, input_shape=(win_length, num_features), 
                                       return_sequences=False))
    else:
        model.add(tf.keras.layers.LSTM(n_nodes, return_sequences=False))

    model.add(tf.keras.layers.Dropout(dropout))
    model.add(tf.keras.layers.Dense(1))
    
    model.compile(loss=tf.losses.MeanSquaredError(),
                  optimizer=tf.optimizers.Adam(learning_rate),
                  metrics=['MAE', 'MSE', 'MAPE'])
    
    model.fit(train_generator, epochs=50,
                validation_data=validate_generator,
                shuffle=False, verbose=2,
                callbacks=[TuneReportCallback({'Mean Squared Error': 'MSE'})])

def alternative_function(config):
    MSE=config['n_nodes']*config['n_layers']
    return {'Mean Squared Error': MSE}

# 4. Define a search space.
search_space = {
    'n_nodes': tune.grid_search([12,15,18,21,24,27,30]),
    'n_layers': tune.grid_search([0,1,2]),
    'learning_rate': tune.grid_search([1e-2,1e-3,1e-4]),
    'dropout': tune.grid_search([0.2,0.3,0.4])
}

asha_scheduler=tune.schedulers.ASHAScheduler(
    max_t=50,
    grace_period=5
    )

# 5. Start a Tune run and print the best result.
if __name__ == "__main__":
    analysis=tune.run(train_model, 
                      config=search_space, 
                      metric="Mean Squared Error", 
                      mode="min", 
                      scheduler=asha_scheduler, 
                      verbose=3,
                      reuse_actors=True)
    print('Best Configuration:', analysis.get_best_config)
    print('Best Trial:', analysis.best_trial)
    print('Logdir of Best Configuration:', analysis.best_logdir)
    
    df_results = analysis.results_df
    df_results.to_csv('MA_PerfectForesight_LSTM_GridSearch_Results.csv')
```

### Anything else

I tried running the code 2 times and got the same error both times (the second time about 45 mins earlier than the first time). I'm running the code on a remote desktop (in my uni) with:
**Processor:** AMD Ryzen Threadripper 2990WX 32-Core Processor   3.00 GHz (64 CPUs)
**RAM:** 128 GB
**GPU:** NVIDIA GeForce GT710

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

**Comment by Numchies** [comment]
Update: The program has been running now for 24 hours after I removed the scheduler and set `resources_per_trial={'cpu': 2}`. Since I have removed the scheduler before and it still crashed, I think me setting the resources per trial somehow improves the code to not give me this RayActorError. Can anyone explain why?

**Comment by dshi3553usyd** [comment]
I got the same problem when I run my ray tune for a simple GCN model.

**Comment by zzNancy1106** [comment]
> 

I met the same problem. Have you solved this problem?

**Comment by LjessonS** [comment]
I also met the same problem

**Comment by peterghaddad** [comment]
I also am having the exact issue on Ray Clusters.

**Comment by peterghaddad** [comment]
@krfricke Know any contributors that may have insight? I am running Ray 1.12.1

**Comment by krfricke** [comment]
The error message:

```
ImportError: DLL load failed while importing _pywrap_tensorflow_internal: Die Auslagerungsdatei ist zu klein, um diesen Vorgang durchzuführen.
```

suggests that your system runs out of memory and tries to swap the memory onto disk. However, the disk is either too small or windows is configured to not allow more than some amount of data to be swapped, and thus the task fails.

The reason for that is concurrency - on a system with 64 CPUs Tune will per default run up to 64 trials in parallel. If each of these trials takes up more than 2 GB of ram, your system ram is exhausted.

That's also the reason why setting `resources_per_trial={'cpu': 2}` solves the problem - you implicitly limit concurrency to 32 trials (as 32 * 2 CPUs = 64 CPUs). Thus every single trial has more memory available.

Instead of explicitly configuring this, you can also use `max_concurrent_trials=32` or `max_concurrent_trials=16` or so in your call to `tune.run()`.

**Comment by xwjiang2010** [comment]
Adding to what @krfricke has said, there is also `max_failures` that you can set that will automatically retry the failed trial from latest checkpoint (make sure that you checkpoint at a reasonable frequency for this to be meaningful).

**Comment by krfricke** [comment]
Closing this for now as it seems to be a resource setup-related issue. Please feel free to re-open if the issue persists.
